bitcoin/src/crypto/key.rs

```diff
@@ -230,13 +227,53 @@ impl fmt::Display for PublicKey {
     }
 }
 
+/// Error returned while constructing public key from string
+#[derive(Debug, PartialEq)]
+pub enum ParsePubKeyError {
+    /// Error originated while parsing string
+    StringError(FromSliceError),
+    /// Hex decoding error.
+    Hex(hex::HexToArrayError),
+    /// `PublicKey` hex should be 66 or 130 digits long.
+    InvalidHexLength(usize),
+}
+
+
+impl From<hex::HexToArrayError> for ParsePubKeyError {
+    fn from(e: hex::HexToArrayError) -> Self { ParsePubKeyError::Hex(e) }
+}
+
+impl fmt::Display for ParsePubKeyError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ParsePubKeyError::*;
+        match *self {
+            StringError(ref e) => write_err!(f, "string error"; e),
+            Hex(ref e) => write_err!(f, "hex error"; e),
+            InvalidHexLength(ref got) => write!(f, "pubkey string should be 66 or 130 digits long, got: {}", got),
+        }
+    }
+}
+
+
 impl FromStr for PublicKey {
-    type Err = Error;
-    fn from_str(s: &str) -> Result<PublicKey, Error> {
+    type Err = ParsePubKeyError;
+    fn from_str(s: &str) -> Result<PublicKey, ParsePubKeyError> {
         match s.len() {
-            66 => PublicKey::from_slice(&<[u8; 33]>::from_hex(s)?),
-            130 => PublicKey::from_slice(&<[u8; 65]>::from_hex(s)?),
-            len => Err(Error::InvalidHexLength(len)),
+            66 => {
+                let res = PublicKey::from_slice(&<[u8; 33]>::from_hex(s)?);
+                match res {
+                    Err(err) => Err(ParsePubKeyError::StringError(err)),
+                    Ok(pubkey) => Ok(pubkey)
+                }
```

Comment:
In 4717cb9a384afc21817a1fba35573311e4ce2b0b:

I think you should use `?` and `From` rather than this manual match. At the very least you can use `.map_err`.

Reply:
I'm surprised `clippy` doesn't complain...

Reply:
Ah sorry this is just bad. I'll fix it.